### ISSUE TITLE
pkg_resources.Requirement.parse() replaces underscores with hyphens -> Cannot include most packages with underscores in UDFs / UDTFs / Stored Procs

### ISSUE THREAD

**Issue by sfc-gh-mgorkow** [snowflakedb/snowpark-python#395]
1. What version of Python are you using?
Python 3.8.13 (default, Mar 28 2022, 11:38:47) 
[GCC 7.5.0]

2. What operating system and processor architecture are you using?
Linux-5.10.25-linuxkit-x86_64-with-glibc2.17

3. What are the component versions in the environment (`pip freeze`)?
`altair==4.2.0
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
asn1crypto==1.4.0
asttokens==2.0.5
attrs==21.4.0
backcall==0.2.0
backports.zoneinfo==0.2.1
beautifulsoup4==4.11.1
bleach==5.0.1
blinker==1.4
blis==0.7.8
cachetools==4.2.2
catalogue==2.0.7
certifi==2021.10.8
cffi==1.15.1
chardet==4.0.0
charset-normalizer==2.0.4
click==8.1.3
click-plugins==1.1.1
cligj==0.7.2
cloudpickle==2.0.0
commonmark==0.9.1
cryptography==36.0.2
cycler==0.11.0
cymem==2.0.6
debugpy==1.6.0
decorator==5.1.1
defusedxml==0.7.1
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.3.0/en_core_web_sm-3.3.0-py3-none-any.whl
entrypoints==0.4
et-xmlfile==1.1.0
executing==0.8.3
fastjsonschema==2.15.3
filelock==3.7.1
Fiona==1.8.21
fonttools==4.33.3
geojson==2.5.0
geopandas==0.9.0
gitdb==4.0.9
GitPython==3.1.27
gmaps==0.9.0
htbuilder==0.6.0
huggingface-hub==0.8.1
idna==3.3
imageio==2.19.3
importlib-metadata==4.12.0
importlib-resources==5.8.0
ipykernel==6.15.0
ipython==8.4.0
ipython-genutils==0.2.0
ipywidgets==7.7.1
iteration-utilities==0.11.0
jedi==0.18.1
Jinja2==3.1.2
joblib==1.1.0
jsonschema==4.6.1
jupyter==1.0.0
jupyter-client==7.3.4
jupyter-console==6.4.4
jupyter-core==4.10.0
jupyterlab-pygments==0.2.2
jupyterlab-widgets==1.1.1
kiwisolver==1.4.3
langcodes==3.3.0
lxml==4.9.0
MarkupSafe==2.1.1
matplotlib==3.5.2
matplotlib-inline==0.1.3
mistune==0.8.4
munch==2.5.0
murmurhash==1.0.7
nbclient==0.6.6
nbconvert==6.5.0
nbformat==5.4.0
nest-asyncio==1.5.5
networkx==2.8.4
nltk==3.7
notebook==6.4.12
numpy==1.22.3
openpyxl==3.0.10
oscrypto==1.2.1
packaging==21.3
pandas==1.4.2
pandas-datareader==0.10.0
pandocfilters==1.5.0
parso==0.8.3
pathy==0.6.2
patsy==0.5.2
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.1.1
preshed==3.0.6
prometheus-client==0.14.1
prompt-toolkit==3.0.30
protobuf==3.20.1
psutil==5.9.1
ptyprocess==0.7.0
pure-eval==0.2.2
pyarrow==6.0.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.21
pycryptodomex==3.15.0
pydantic==1.8.2
pydeck==0.7.1
Pygments==2.12.0
PyJWT==2.1.0
Pympler==1.0.1
pyOpenSSL==22.0.0
pyparsing==3.0.9
pyproj==3.3.1
pyrsistent==0.18.1
python-dateutil==2.8.2
pytz==2022.1
pytz-deprecation-shim==0.1.0.post0
PyWavelets==1.3.0
PyYAML==6.0
pyzmq==23.2.0
qtconsole==5.3.1
QtPy==2.1.0
regex==2021.8.3
requests==2.28.1
rich==12.4.4
sacremoses==0.0.53
scikit-image==0.19.2
scikit-learn==1.0.2
scipy==1.7.1
seaborn==0.9.0
semver==2.13.0
Send2Trash==1.8.0
sentence-transformers==2.2.2
sentencepiece==0.1.95
Shapely==1.8.2
six==1.16.0
smart-open==5.2.1
smmap==5.0.0
snowflake-connector-python==2.7.9
snowflake-snowpark-python==0.7.0
soupsieve==2.3.2.post1
spacy==3.3.0
spacy-legacy==3.0.9
spacy-loggers==1.0.2
sqlparse==0.4.2
srsly==2.4.3
st-annotated-text==3.0.0
stack-data==0.3.0
statsmodels==0.13.2
streamlit==1.10.0
terminado==0.15.0
thinc==8.0.17
threadpoolctl==3.1.0
tifffile==2022.5.4
timm==0.5.4
tinycss2==1.1.1
tokenizers==0.12.1
toml==0.10.2
toolz==0.11.2
torch==1.8.1
torchvision==0.9.1
tornado==6.1
tqdm==4.64.0
traitlets==5.3.0
transformers==4.18.0
typer==0.4.1
typing_extensions==4.1.0
tzdata==2022.1
tzlocal==4.2
urllib3==1.26.8
validators==0.20.0
wasabi==0.9.1
watchdog==2.1.9
wcwidth==0.2.5
webencodings==0.5.1
widgetsnbextension==3.6.1
wordcloud==1.8.2.2
xgboost==1.5.0
zipp==3.8.0`

4. What did you do?
I tried to create a UDF with dependencies on a package with underscores.
`def test() -> None:
    import spacy
    self.nlp = spacy.load("en_core_web_sm")
    return
session.udf.register(func=test, packages=['spacy','spacy-model-en_core_web_sm'])`

5. What did you expect to see?
Expected Result: UDF is created
Result: Error-Message "Cannot add package spacy-model-en-core-web-sm because it is not available in Snowflake."

**My investigations so far:**
As you can see, the resulting error message replaced all underscores with hyphens.
My investigation showed that the problem comes from the [pkg_resources.Requirement.parse](https://github.com/snowflakedb/snowpark-python/blob/e0cc59f81bc48e027223528d112825edc9dd0b65/src/snowflake/snowpark/session.py#L708) function which replaces underscores with hyphens.
As a result, the package_name does not match anymore.

I can replicate this behaviour by running the following code:
`
import pkg_resources
package_req = pkg_resources.Requirement.parse('spacy-model-en_core_web_sm')
package_name = package_req.key
print(package_name)
`

Output: spacy-model-en-core-web-sm

Using underscores in package_names is also discouraged as described [here](https://peps.python.org/pep-0008/#package-and-module-names).
> Python packages should also have short, all-lowercase names, although the use of underscores is discouraged.

List of currently affected packages:
`async_generator
backports.functools_lru_cache
backports.shutil_get_terminal_size
backports_abc
category_encoders
cmake_setuptools
console_shortcut
cx_oracle
essential_generators
et_xmlfile
factory_boy
flask_cors
get_terminal_size
huggingface_hub
idna_ssl
importlib_resources
ipython_genutils
jupyter_client
jupyter_core
jupyter_server
jupyter_telemetry
jupyterlab_pygments
jupyterlab_server
jupyterlab_widgets
korean_lunar_calendar
line_profiler
memory_profiler
mpl_sample_data
multi_key_dict
mypy_extensions
opentracing_instrumentation
opt_einsum
plaster_pastedeploy
posix_ipc
prometheus_client
pure_eval
pyramid_jinja2
python_http_client
pyviz_comms
readme_renderer
requests_download
requests_ntlm
ruamel_yaml
semantic_version
service_identity
smart_open
spacy-model-en_core_web_lg
spacy-model-en_core_web_md
spacy-model-en_core_web_sm
spacy-model-en_core_web_trf
stack_data
typing_inspect
vega_datasets
win32_setctime`

**Comment by sfc-gh-jdu** [comment]
@sfc-gh-mgorkow Thanks for your feedback. We actually already realized this issue, and handle some popular packages with underscores [here](https://github.com/snowflakedb/snowpark-python/blob/e0cc59f81bc48e027223528d112825edc9dd0b65/src/snowflake/snowpark/_internal/utils.py#L64-L91). `spacy-model-en_core_web_sm` is not included, so would you like to add this one into this dict?

**Comment by sfc-gh-mgorkow** [comment]
@sfc-gh-jdu This mapping is only applied if I add the package as a module, not by it's name (string-type). So it wouldn't change anything in my case.

However, even if I add `"spacy-model-en_core_web_sm":"en_core_web_sm"` to this dictionary and try to include the package via `import en_core_web_sm; session.add_packages([en_core_web_sm]);` I receive the following error-message: `The 'spacy-model-en_core_web_sm' distribution was not found and is required by the application`

I tried the same with `import xgboost; session.add_packages([xgboost]);` which is already in the mapping -> same error message.
At this point I don't even understand what this mapping does ... and it doesn't feel like it's a consistent solution.

I don't know yet but I hope there is a better solution...

**Comment by sfc-gh-jdu** [comment]
> en_core_web_sm



> @sfc-gh-jdu This mapping is only applied if I add the package as a module, not by it's name (string-type). So it wouldn't change anything in my case.
> 
> However, even if I add `"spacy-model-en_core_web_sm":"en_core_web_sm"` to this dictionary and try to include the package via `import en_core_web_sm; session.add_packages([en_core_web_sm]);` I receive the following error-message: `The 'spacy-model-en_core_web_sm' distribution was not found and is required by the application`
> 
> I tried the same with `import xgboost; session.add_packages([xgboost]);` which is already in the mapping -> same error message. At this point I don't even understand what this mapping does ... and it doesn't feel like it's a consistent solution.
> 
> I don't know yet but I hope there is a better solution...

Hmm it (adding module directly) turns out being a bug in our code, and we'll fix it in the next release. For passing package as a string, yea it looks like a general issue and I'll discuss with the team. Thanks for the feedback!
